### PR TITLE
Support flask-caching with the Flask-Cache tracer

### LINKED ISSUE #2242: Make flask-cache tracer work with the flask-caching fork
The documentation mentions that there's [tracing is available for "Flask Cache"](https://ddtrace.readthedocs.io/en/stable/integrations.html#module-ddtrace.contrib.flask_cache). This project has been unmaintained for 7 years however: https://github.com/thadeusb/flask-cache

There is a maintained fork at https://github.com/sh4nks/flask-caching, and even though it's API-compatible with the old library, ddtrace is not compatible because [the import path is different](https://github.com/sh4nks/flask-caching#compatibility-with-flask-cache) and the dd-trace module [hardcodes the import path of the old library](https://github.com/DataDog/dd-trace-py/blob/dea3a6435612293aa592997e87983f3e58b75947/ddtrace/contrib/flask_cache/tracers.py#L7).

### PR DESCRIPTION
**Pull request by nizox**
## Description
[flask-caching](https://flask-caching.readthedocs.io/en/latest/) is a fork of [Flask-Cache](https://github.com/thadeusb/flask-cache) with minor incompatibilities. This PR adds support for flask-caching with the same tracer than Flask-Cache.

The `get_traced_cache` function is exposed when flask-caching or Flask-Cache is installed. A detection mechanism is added to use the relevant Cache class. Flask-Cache has priority over flask-caching in order to keep compatibility with the current tracer behaviour.

Also, the `get_traced_cache` function gains a new keyword argument called `cache_cls` to override this detection.

Resolves #2242.

## Checklist
- [ ] Added to the correct milestone.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: ddtrace/contrib/flask_cache/tracers.py [hunk 1]
```diff
@@ -41,7 +55,7 @@ class TracedCache(Cache):
         _datadog_service = service
         _datadog_meta = meta
 
-        def __trace(self, cmd):
+        def __trace(self, cmd, write=False):
```
Comment (Yun-Kim) [inline ddtrace/contrib/flask_cache/tracers.py:58]: Where is `write=False` used?

Comment (nizox) [reply ddtrace/contrib/flask_cache/tracers.py:58]: Indeed it lacks documentation, but the redis backend of Flask-Cache was modified upstream to support multiple connections, one for writing and one for reading. This argument selects the right connection.

**Review by brettlangdon — COMMENTED** [review]
we just need to update docs to list that we support flask-caching and min-version we support, otherwise lgtm

File: ddtrace/contrib/flask_cache/tracers.py (hunk 1 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/flask_cache/tracers.py:58]: is this comment stale @nizox? looks like the `write` argument isn't used in this function 🤔

File: ddtrace/contrib/flask_cache/tracers.py [hunk 2]
```diff
@@ -54,9 +68,10 @@ def __trace(self, cmd):
             # set analytics sample rate
             s.set_tag(ANALYTICS_SAMPLE_RATE_KEY, config.flask_cache.get_analytics_sample_rate())
             # add connection meta if there is one
-            if getattr(self.cache, "_client", None):
+            client = _extract_client(self.cache)
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/flask_cache/tracers.py:71]:
```suggestion
            client = _extract_client(self.cache, write)
```
?

Comment (nizox) [reply ddtrace/contrib/flask_cache/tracers.py]: Good catch! The argument is used only for the Redis Sentinel backend which is unfortunately untested by our test suite and so it didn't catch the issue. I tried to add a Redis Sentinel test but it's not trivial and requires setting up multiple instances of Redis.

File: ddtrace/contrib/flask_cache/tracers.py (hunk 1 above)
Comment (nizox) [reply ddtrace/contrib/flask_cache/tracers.py:58]: Good catch! I updated the PR.

**Review by brettlangdon — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Pull request merged**